LibWeb: Correctly lay out block elements inside inline elements

gmta opened the pull request:
The contents of inline elements containing a block element were incorrectly ordered, e.g.

```html
<b>foo<div>bar</div>baz</b>
```

...did not render the `<div>` as being in between the `foo` and `baz` fragments.

| Before | After |
|---|---|
| ![Screenshot_20250116_164644](https://github.com/user-attachments/assets/34c32cb8-e7d1-48e7-a1be-693b48fa9373) | ![image](https://github.com/user-attachments/assets/5f53cf9f-23c6-40c7-aa9c-d61f8970ec74) |


Excerpt from `Web::Layout::TreeBuilder`:

---

Block nodes inside inline nodes are allowed, but to maintain the invariant that either all layout children are
inline or non-inline, we need to rearrange the tree a bit. All inline ancestors up to the node we've inserted are
wrapped in an anonymous block, which is inserted into the nearest non-inline ancestor. We then recreate the inline
ancestors in another anonymous block inserted after the node so we can continue adding children.

Effectively, we try to turn this:

    InlineNode 1
      TextNode 1
      InlineNode N
        TextNode N
        BlockContainer (node)

Into this:

    BlockContainer (anonymous "before")
      InlineNode 1
        TextNode 1
        InlineNode N
          TextNode N
    BlockContainer (anonymous "middle") continuation
      BlockContainer (node)
    BlockContainer (anonymous "after")
      InlineNode 1 continuation
        InlineNode N

To be able to reconstruct their relation after restructuring, layout nodes keep track of their continuation. The
top-most inline node of the "after" wrapper points to the "middle" wrapper, which points to the top-most inline node
of the "before" wrapper. All other inline nodes in the "after" wrapper point to their counterparts in the "before"
wrapper, to make it easier to create the right paintables since a DOM::Node only has a single Layout::Node.

Appending then continues in the "after" tree. If a new block node is then inserted, we can reuse the "middle" wrapper
if no inline siblings exist for node or its ancestors, and leave the existing "after" wrapper alone. Otherwise, we
create new wrappers and extend the continuation chain.

Inspired by: https://webkit.org/blog/115/webcore-rendering-ii-blocks-and-inlines/

kalenikaliaksandr reviewed (commented):
have you considered solving the problem on paintable tree level? I suppose webkit has to use continuations on layout tree because they do not have paintable tree concept, while we do and support 1:N relationship between layout and paintable nodes.

gmta commented:
@kalenikaliaksandr @awesomekling I've rearranged some things:

- Determining the united border/padding/margin/content rects is now the responsibility of `PaintableBox` again.
- The continuation chain is still only part of the layout tree however; it's not clear to me if we're always going to have paintables for parts of split-up inline elements that might not be visible (e.g. in `<b><div>foo</div>bar</b>`) and if so, how we're going to reliably link up the right paintables during creation of the painting tree. I've left a FIXME in `PaintableBox` to this end.
- More DOM methods were updated and I've added a test for offsetTop/Left/Width/Height.

kalenikaliaksandr reviewed (changes requested):
could we add a test for hit-testing just so we can be sure it works as expected when it comes to continuations?

gmta commented:
> could we add a test for hit-testing just so we can be sure it works as expected when it comes to continuations?

Absolutely, in fact, I'm also working on a bugfix for something depending on this that involves hit-testing. I'll tack on the commit to this PR.

gmta commented:
> Ran into a simple case that doesn't work correctly:
> 
> ```
> Well <span> hello <div>friends
> ```

Well that's embarrassing! I'll take a look

awesomekling commented:
And here's a basic dynamic case that doesn't work -- looks like we don't propagate style changes to the split layout nodes.

```html
<!doctype html>
<style>
* { font-family: sans-serif; }
</style>
<script>
    onload = function() {
        document.body.offsetWidth; // force a layout
        a.setAttribute("style", null);
    }
</script>
</head>
<body>
    <span id="a" style="font-weight: bold;">hello <div>inner</div></span>
</body>
```

gmta commented:
@kalenikaliaksandr @awesomekling some more updates:

* Added a new test performing hit tests on parts of an inline continuation chain;
* Fixed previous inline siblings not being moved to the "before" wrapper (`foo<b>bar<div>baz`)
* Fixed dynamic style updates not being propagated along the continuation chain

W/ regard to that last point, it seems to be working correctly but similar to propagating styles to anonymous wrappers (see the `FIXME` in that method), we might benefit from having shared computed styles at some point.

kalenikaliaksandr commented on Libraries/LibWeb/Painting/PaintableBox.cpp:
```diff
@@ -214,13 +214,86 @@ CSSPixelRect PaintableBox::compute_absolute_paint_rect() const
+    return result.value_or({});
+}
+
+CSSPixelRect PaintableBox::absolute_united_border_box_rect() const
+{
+    return united_rect_for_continuation_chain(*this, [](auto const& paintable_box) {
+        return paintable_box.absolute_border_box_rect();
+    });
+}
```
hmm, shouldn't `absolute_border_box_rect()` behave like `absolute_united_border_box_rect()`. I would expect anything that does `dom_node->paintable_box()->absolute_border_box_rect()` to want united box by default.

gmta replied:
I've considered that, but that's not what you always want. Consider this fragment:

```html
foo<b>bar<div>baz</div></b>
```

"foo" should not be part of the paintable's rect, only "bar" and the `div` should be. At the moment I think the main purpose for the united versions of the rects is the inspector overlay; hit testing still requires the smaller rects.

gmta commented:
Hit-testing was improved with respect to continuation chains, where we now consider inline elements that are parents of block elements to be valid targets of a hit-test even if that block element has `pointer-events: none` set.

This fixes the `<a><figure /></a>` links in use on the homepage of https://tweakers.net:

![image](https://github.com/user-attachments/assets/46a8ffb9-5220-4028-bf6a-1fb95fb00ae8)

kalenikaliaksandr commented on Libraries/LibWeb/Painting/PaintableBox.cpp:
```diff
@@ -978,18 +975,37 @@ Optional<HitTestResult> PaintableBox::hit_test(CSSPixelPoint position, HitTestTy
+        result->index_in_node = result->paintable->dom_node() ? result->paintable->dom_node()->index() : 0;
+        result->paintable = result->paintable->parent();
+
+        // If the new parent is part of a continuation, we need to follow the chain to the original inline node.
```
could you add a bit more clarification here why we can't stop on a paintable created by continuation and need to follow up until the origin node?

gmta replied:
> could you add a bit more clarification

Absolutely, expanded the second comment in that piece of code.

> why we can't stop on a paintable created by continuation

We could actually be more specific: if we end up on an anonymous wrapper that is part of a continuation chain, this is always a "middle" part (since the "before" and "after" parts are inline nodes) and we need to follow the chain all the way to the "before" part to find the actual parent node. Otherwise, we would return a hit result on the wrapper.

circl-lastname commented:
Fixes #959 (?)
